README
NAME
    Dist::Inkt - yet another distribution builder

STATUS
    Experimental.

DESCRIPTION
    Dist::Zilla didn't have the prerequisite amount of crazy for me, so I
    wrote this instead.

    Dist::Inkt itself does virtually nothing; it creates an empty directory,
    generates a MANIFEST file, and then wraps it all up into a gzipped
    tarball. But it provides various hooks along the way for subclasses to
    grab hold of. So the general idea is that you write a subclass of
    Dist::Inkt, which consumes various Moose::Roles to do the actual work of
    populating the distribution with files.

    As such, Dist::Inkt is not so much a distribution builder, as it is a
    framework for writing your own distribution builder.

    Several roles of varying utility are bundled with Dist::Inkt, as is
    Dist::Inkt::Profile::TOBYINK, a subclass of Dist::Inkt which consumes all
    of these roles.

COMPANIONS
    Dist::Inkt does just one thing - building the tarball from some checkout
    of the repo.

    Although roles could theoretically be written for other tasks, out of the
    box, Dist::Inkt doesn't do any of the following:

    Minting new distributions
        I'm writing a separate tool, Dist::Inktly::Minty for that.

    Test suite running
        Use App::Prove or App::ForkProve.

    CPAN Uploading
        Use CPAN::Uploader.

    Changing the version number across many files
        Use Perl::Version.

    Integration with version control tools
        Just use `hg` or `svn` or `git` of whatever as you normally would.
        None of the files generated by Dist::Inkt should probably be checked
        into your repo.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=Dist-Inkt>.

SEE ALSO
    If you are not me, then you may well want one of these instead:

    *   Dist::Zilla

    *   Dist::Milla

    *   Minilla

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2013 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Dist-Inkt
=========

Created:      2013-08-07
Home page:    <https://metacpan.org/release/Dist-Inkt>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Dist-Inkt>
Wiki:         <https://github.com/tobyink/p5-dist-inkt/wiki>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.023	2014-09-17

 - Avoid outputting anonymous class names.
 - Improve command-line option handling.

0.022	2014-06-19

 [ Bug Fixes ]
 - Fix check for existence of meta/META.PL.

 [ Other ]
 - Filter $meta->{provides} using $meta->{no_index}.

0.021	2014-06-19

 [ Documentation ]
 - Fix documentation of the `--TRAVIS` option.

 [ Other ]
 - Generated Makefile.PL should cope better if CPAN::Meta::Requirements is
   missing.

0.020	2014-06-16

 [ Documentation ]
 - Added stub pod to command line scripts.

 [ Other ]
 - Better integration with haarg's perl-travis-helper stuff.
 - Improvements to command line scripts.

0.019	2014-06-15

 [ Packaging ]
 - Add missing Path::FindDev dependency.

 [ Other ]
 - Cope when ~/perl5/standard-documents/ is missing.
 - Support PERL_DIST_INKT_NOSIGNATURE environment variable to skip building
   a SIGNATURE file.

0.018	2014-06-14

 [ Packaging ]
 - Add missing File::chdir dependency.

0.017	2014-05-26

 - Simplify usage of Dist::Inkt::Profile::Simple, and example.

0.016	2014-05-26

 [ Packaging ]
 - Split out a lot of functionality into Dist-Inkt-DOAP.

 [ Other ]
 - Added: Dist::Inkt::Profile::Simple.
 - Added: Dist::Inkt::Role::CPANfile.
 - Added: Dist::Inkt::Role::WriteDOAPLite.

0.015	2014-03-10

 - Guess repository details from repository web URL for github.com.

0.014	2014-01-21

 - Added: Bundle a `distinkt-travisyml` script to automate creating
   .travis.yml files.
 - Generated Makefile.PL explicitly lists test suite directories.
 - Use Path::Tiny's `subsumes` method over home-brew version.

0.013	2013-09-22

 - Generated Makefile.PL performs a better check for minimum Perl version.

0.012	2013-09-19

 [ Bug Fixes ]
 - Generating LICENSE files now understands CPAN::Meta::Spec 2.0 license
   codes.

0.011	2013-09-19

 [ Bug Fixes ]
 - META.json/yml license codes now conform to CPAN::Meta::Spec 2.0 instead
   of META-spec 1.4.

0.010	2013-09-11

 - Added: Output an INSTALL file.
 - Added: Output optional_features in META.json.
 - Added: Use dynamic config in Makefile.PL to add dependencies for
   optional features.
 - Updated: Support `feature` property from DOAP deps.

0.009	2013-09-06

 [ Bug Fixes ]
 - Correct %script to %scripts in Dist::Inkt::Role::MetaProvidesScripts.
 - Make sure EUMM $VERSION gets evaled.
   Graham Knop++
 - Replace => with = in assignment in Makefile.PL.
   Graham Knop++

0.008	2013-09-06

 [ Packaging ]
 - Depend on a newer version of RDF::DOAP, thus we get more awesome Changes
   files.

 [ Other ]
 - Add `x_breaks` key to META.json if appropriate.
 - Allow `source_for_readme` to be set in Dist::Inkt constructor (and thus
   in dist.ini).
 - Don't add `x_provides_scripts` to META.json unless scripts are actually
   provided.
 - Perform conflicts checking in Makefile.PL.

0.007	2013-08-27

 - Rename META.ttl -> doap.ttl.

0.006	2013-08-13

 - Guess some copyrights for the COPYRIGHT file.

0.005	2013-08-13

 - Added: Support the `x_help_wanted` key for META.json/META.yml.
 - Updated: Use the shiny, new RDF::DOAP for a lot of the behind-the-scenes
   DOAP processing.

0.004	2013-08-10

 - Attempt to support ExtUtils::MakeMaker 6.17. Don't pass WriteMakefile a
   LICENSE unless EUMM >= 6.3001; don't pass it a DISTVNAME at all.
 - For ExtUtils::MakeMaker versions that don't support BUILD_REQUIRES,
   bundle it in with PREREQ_PM, not CONFIGURE_REQUIRES.
 - For ExtUtils::MakeMaker versions that don't support TEST_REQUIRES,
   bundle it in with BUILD_REQUIRES, not CONFIGURE_REQUIRES.

0.003	2013-08-10

 - Don't pass ExtUtils::MakeMaker the AUTHOR as an arrayref unless its
   version number >= 6.5702.

0.002	2013-08-09

 - Add shebang to `bin/distinkt-dist`.
 - Avoid creating empty 'x_contributors' keys in META.json/META.yml.

0.001	2013-08-09	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"           => "yet another distribution builder",
  "author"             => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config"     => 0,
  "generated_by"       => "Dist::Inkt::Profile::TOBYINK version 0.022, CPAN::Meta::Converter version 2.140640",
  "keywords"           => ["CPAN", "Distribution", "Package", "Perl", "Release"],
  "license"            => ["perl_5"],
  "meta-spec"          => {
                            url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                            version => 2,
                          },
  "name"               => "Dist-Inkt",
  "no_index"           => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"            => {
                            configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                            runtime => {
                              requires => {
                                "CPAN::Changes"            => 0,
                                "CPAN::Meta"               => 0,
                                "CPAN::Meta::Requirements" => 0,
                                "Data::Dump"               => 0,
                                "File::chdir"              => 0,
                                "List::MoreUtils"          => 0,
                                "Module::CPANfile"         => 1.0002,
                                "Module::Metadata"         => 0,
                                "Module::Runtime"          => 0,
                                "Module::Signature"        => 0,
                                "Moose"                    => "2.0800",
                                "namespace::autoclean"     => 0,
                                "Path::FindDev"            => 0,
                                "Path::Iterator::Rule"     => 0,
                                "Path::Tiny"               => 0,
                                "perl"                     => 5.010001,
                                "RDF::DOAP::Lite"          => 0,
                                "Software::License"        => 0,
                                "Types::Path::Tiny"        => 0,
                                "Types::Standard"          => 0,
                              },
                            },
                            test => { requires => { "Test::More" => 0.96 } },
                          },
  "provides"           => {
                            "Dist::Inkt"                              => { file => "lib/Dist/Inkt.pm", version => 0.023 },
                            "Dist::Inkt::Profile::Simple"             => { file => "lib/Dist/Inkt/Profile/Simple.pm", version => 0.023 },
                            "Dist::Inkt::Role::CopyFiles"             => { file => "lib/Dist/Inkt/Role/CopyFiles.pm", version => 0.023 },
                            "Dist::Inkt::Role::CopyStandardDocuments" => {
                                                                           file => "lib/Dist/Inkt/Role/CopyStandardDocuments.pm",
                                                                           version => 0.023,
                                                                         },
                            "Dist::Inkt::Role::CPANfile"              => { file => "lib/Dist/Inkt/Role/CPANfile.pm", version => 0.023 },
                            "Dist::Inkt::Role::MetaProvides"          => { file => "lib/Dist/Inkt/Role/MetaProvides.pm", version => 0.023 },
                            "Dist::Inkt::Role::MetaProvidesScripts"   => {
                                                                           file => "lib/Dist/Inkt/Role/MetaProvidesScripts.pm",
                                                                           version => 0.023,
                                                                         },
                            "Dist::Inkt::Role::SignDistribution"      => { file => "lib/Dist/Inkt/Role/SignDistribution.pm", version => 0.023 },
                            "Dist::Inkt::Role::WriteDOAPLite"         => { file => "lib/Dist/Inkt/Role/WriteDOAPLite.pm", version => 0.023 },
                            "Dist::Inkt::Role::WriteINSTALL"          => { file => "lib/Dist/Inkt/Role/WriteINSTALL.pm", version => 0.023 },
                            "Dist::Inkt::Role::WriteLICENSE"          => { file => "lib/Dist/Inkt/Role/WriteLICENSE.pm", version => 0.023 },
                            "Dist::Inkt::Role::WriteMakefilePL"       => { file => "lib/Dist/Inkt/Role/WriteMakefilePL.pm", version => 0.023 },
                            "Dist::Inkt::Role::WriteMetaJSON"         => { file => "lib/Dist/Inkt/Role/WriteMetaJSON.pm", version => 0.023 },
                            "Dist::Inkt::Role::WriteMetaYML"          => { file => "lib/Dist/Inkt/Role/WriteMetaYML.pm", version => 0.023 },
                            "Dist::Inkt::Role::WriteREADME"           => { file => "lib/Dist/Inkt/Role/WriteREADME.pm", version => 0.023 },
                          },
  "release_status"     => "stable",
  "resources"          => {
                            bugtracker   => { web => "http://rt.cpan.org/Dist/Display.html?Queue=Dist-Inkt" },
                            homepage     => "https://metacpan.org/release/Dist-Inkt",
                            license      => ["http://dev.perl.org/licenses/"],
                            repository   => {
                                              type => "git",
                                              url  => "git://github.com/tobyink/p5-dist-inkt.git",
                                              web  => "https://github.com/tobyink/p5-dist-inkt",
                                            },
                            x_identifier => "http://purl.org/NET/cpan-uri/dist/Dist-Inkt/project",
                            x_wiki       => "https://github.com/tobyink/p5-dist-inkt/wiki",
                          },
  "version"            => 0.023,
  "x_help_wanted"      => ["tester", "documenter"],
  "x_provides_scripts" => {
                            "distinkt-dist"      => { file => "bin/distinkt-dist" },
                            "distinkt-travisyml" => { file => "bin/distinkt-travisyml" },
                          },
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

